src/shared/components/catalog/CatalogController.tsx

```diff
@@ -0,0 +1,205 @@
+import {
+  CatalogCategory,
+  CatalogFilters,
+  CatalogQueryParams,
+  CatalogStringMap,
+  CatalogType,
+  CatalogItem,
+  CatalogItemAttribute,
+  CatalogService,
+} from './utils/types';
+import { RenderCell } from '../virtualized-grid/types';
+import PageHeading from '../headings/PageHeading';
+
+type CatalogControllerProps = CatalogService & {
+  enableDetailsPanel?: boolean;
+  hideSidebar?: boolean;
+  title: string;
+  description: string;
+  categories?: CatalogCategory[];
+};
+
+const CatalogController: React.FC<CatalogControllerProps> = ({
+  type,
+  items,
+  itemsMap,
+  loaded,
+  loadError,
+  catalogExtensions,
+  enableDetailsPanel,
+  title: defaultTitle,
+  description: defaultDescription,
+  hideSidebar,
+  categories,
+}) => {
+  const { t } = useTranslation();
+  const { pathname } = useLocation();
+  const queryParams = useQueryParams();
+
+  const typeExtension = React.useMemo(
+    () => catalogExtensions?.find((extension) => extension.properties.type === type),
+    [catalogExtensions, type],
+  );
+
+  const title = typeExtension?.properties.title ?? defaultTitle;
+  const description = typeExtension?.properties.catalogDescription ?? defaultDescription;
+
+  const filterGroups: string[] = React.useMemo(() => {
+    return (
+      typeExtension?.properties.filters?.map((filter: CatalogItemAttribute) => filter.attribute) ??
+      []
+    );
+  }, [typeExtension]);
+
+  const filterGroupNameMap: CatalogStringMap = React.useMemo(() => {
+    return (
+      typeExtension?.properties.filters?.reduce((map, filter: CatalogItemAttribute) => {
+        map[filter.attribute] = filter.label;
+        return map;
+      }, {}) ?? {}
+    );
+  }, [typeExtension]);
+
+  const groupings: CatalogStringMap = React.useMemo(() => {
+    return (
+      typeExtension?.properties.groupings?.reduce((map, group: CatalogItemAttribute) => {
+        map[group.attribute] = group.label;
+        return map;
+      }, {}) ?? {}
+    );
+  }, [typeExtension]);
+
+  const breadcrumbs = React.useMemo(() => {
+    const categoryParam = queryParams.get(CatalogQueryParams.CATEGORY);
+    const keywordParam = queryParams.get(CatalogQueryParams.KEYWORD);
+    const sortParam = queryParams.get(CatalogQueryParams.SORT_ORDER);
+    const params = new URLSearchParams({
+      ...(categoryParam ? { [CatalogQueryParams.CATEGORY]: categoryParam } : {}),
+      ...(keywordParam ? { [CatalogQueryParams.KEYWORD]: keywordParam } : {}),
+      ...(sortParam ? { [CatalogQueryParams.SORT_ORDER]: sortParam } : {}),
+    });
+    const crumbs = [
+      {
+        name: t('appstudio~Samples Catalog'),
+        path: `${pathname}?${params.toString()}`,
+      },
+    ];
+
+    if (type) {
+      crumbs.push({
+        name: title,
+        path: `${pathname}?${CatalogQueryParams.TYPE}=${type}`,
+      });
+    }
+
+    return crumbs;
+  }, [pathname, t, queryParams, title, type]);
+
+  // const selectedItem = React.useMemo(() => {
```

Comment:
Why is this commented out?

Reply:
Discussed offline. See https://github.com/openshift/hac-dev/pull/10#issuecomment-967462839.